File: examples/cli/go/provider.go
package main

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"time"

	slop "github.com/slop-ai/slop-go"
)

const windowSize = 25

func floatPtr(f float64) *float64 { return &f }
func intPtr(i int) *int           { return &i }

func setupProvider(store *Store) *slop.Server {
	server := slop.NewServer("tsk", "tsk")

	// --- search affordance on tasks collection ---
	server.HandleWith("tasks", "search", slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
		query := p.String("query")
		results, err := store.Search(query)
		if err != nil {
			return nil, err
		}
		items := make([]map[string]any, len(results))
		for i, t := range results {
			items[i] = map[string]any{
				"id":    t.ID,
				"title": t.Title,
				"done":  t.Done,
				"due":   t.Due,
				"tags":  t.Tags,
			}
		}
		return map[string]any{"results": items, "count": len(items)}, nil
	}), slop.ActionOpts{
		Label:       "Search tasks",
		Description: "Search tasks by title or tag",
		Idempotent:  true,
		Estimate:    "instant",
		Params:      map[string]string{"query": "string"},
	})

	// --- static user context ---
	server.RegisterFunc("user", func() slop.Node {
		total, done, _, _, _ := store.Stats()
		return slop.Node{
			Type: "context",
			Props: slop.Props{
				"file":        store.path,
				"total_tasks": total,
				"total_done":  done,
			},
		}
	})

	// --- dynamic tasks collection ---
	server.RegisterFunc("tasks", func() slop.Node {
		return buildTasksNode(store)
	})

	// --- collection-level actions registered separately ---
	server.HandleWith("tasks", "add", slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
		title := p.String("title")
		if title == "" {
			return nil, fmt.Errorf("title is required")
		}
		due := p.String("due")
		tags := p.String("tags")
		t, err := store.Add(title, due, tags)
		if err != nil {
			return nil, err
		}
		return map[string]any{"id": t.ID}, nil
	}), slop.ActionOpts{
		Label:    "Add task",
		Estimate: "instant",
		Params:   map[string]string{"title": "string", "due": "string", "tags": "string"},
	})

	server.HandleWith("tasks", "clear_done", slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
		count, err := store.ClearDone()
		if err != nil {
			return nil, err
		}
		return map[string]any{"cleared": count}, nil
	}), slop.ActionOpts{
		Label:       "Clear completed",
		Description: "Remove all completed tasks",
		Dangerous:   true,
		Estimate:    "instant",
	})

	server.HandleWith("tasks", "export", slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
		format := p.String("format")
		if format == "" {
			format = "json"
		}
		// Simulate async: return accepted, then do the work.
		go func() {
			time.Sleep(500 * time.Millisecond)
			_, _ = store.Export(format)
			server.Refresh()
		}()
		return map[string]any{"__async": true, "status": "accepted"}, nil
	}), slop.ActionOpts{
		Label:       "Export tasks",
		Description: "Export tasks to a file",
		Estimate:    "slow",
		Params:      map[string]string{"format": "string"},
	})

	// --- tags collection ---
	server.RegisterFunc("tags", func() slop.Node {
		return buildTagsNode(store)
	})

	server.HandleWith("tags", "rename", slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
		oldName := p.String("old")
		newName := p.String("new")
		if oldName == "" || newName == "" {
			return nil, fmt.Errorf("old and new are required")
		}
		tasks, err := store.All()
		if err != nil {
			return nil, err
		}
		count := 0
		for _, t := range tasks {
			for i, tag := range t.Tags {
				if tag == oldName {
					t.Tags[i] = newName
					count++
					if _, err := store.Edit(t.ID, "", "", strings.Join(t.Tags, ",")); err != nil {
						return nil, err
					}
					break
				}
			}
		}
		return map[string]any{"renamed": count}, nil
	}), slop.ActionOpts{
		Label:    "Rename tag",
		Estimate: "instant",
		Params:   map[string]string{"old": "string", "new": "string"},
	})

	return server
}

func buildTasksNode(store *Store) slop.Node {
	tasks, err := store.SortedBySalience()
	if err != nil {
		return slop.Node{Type: "collection", Summary: "error loading tasks"}
	}

	total, done, pending, overdue, _ := store.Stats()
	now := today()

	// Window the tasks
	windowEnd := windowSize
	if windowEnd > len(tasks) {
		windowEnd = len(tasks)
	}
	windowed := tasks[:windowEnd]

	items := make([]slop.Item, len(windowed))
	for i, t := range windowed {
		items[i] = buildTaskItem(store, t, now)
	}

	node := slop.Node{
		Type: "collection",
		Props: slop.Props{
			"count":   total,
			"pending": pending,
			"overdue": overdue,
		},
		Summary: fmt.Sprintf("%d tasks: %d pending, %d done, %d overdue", total, pending, done, overdue),
		Items:   items,
		Meta: &slop.Meta{
			TotalChildren: intPtr(total),
			Window:        &[2]int{0, windowEnd},
		},
	}

	return node
}

func buildTaskItem(store *Store, t Task, now time.Time) slop.Item {
	sal := taskSalience(t, now)
	urg := taskUrgency(t, now)
	reason := taskReason(t, now)

	props := slop.Props{
		"title": t.Title,
		"done":  t.Done,
	}
	if t.Due != "" {
		props["due"] = t.Due
	}
	if len(t.Tags) > 0 {
		props["tags"] = t.Tags
	}
	if t.Done && t.CompletedAt != "" {
		props["completed_at"] = t.CompletedAt
	}

	// Content ref for notes
	contentRef := map[string]any{
		"type": "text",
		"mime": "text/plain",
	}
	if t.Notes != "" {
		contentRef["summary"] = fmt.Sprintf("%d lines of notes", strings.Count(t.Notes, "\n")+1)
		size := len(t.Notes)
		contentRef["size"] = size
		preview := t.Notes
		if len(preview) > 100 {
			preview = preview[:97] + "..."
		}
		contentRef["preview"] = preview
	} else {
		contentRef["summary"] = "No notes"
	}
	props["content_ref"] = contentRef

	// Actions depend on done state
	actions := slop.Actions{}
	taskID := t.ID

	if t.Done {
		actions["undo"] = slop.WithOpts(
			slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
				_, err := store.Undo(taskID)
				return nil, err
			}),
			slop.ActionOpts{Label: "Mark incomplete", Estimate: "instant"},
		)
	} else {
		actions["done"] = slop.WithOpts(
			slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
				_, err := store.Done(taskID)
				return nil, err
			}),
			slop.ActionOpts{Label: "Complete task", Estimate: "instant"},
		)
		actions["edit"] = slop.WithOpts(
			slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
				title := p.String("title")
				due := p.String("due")
				tags := p.String("tags")
				_, err := store.Edit(taskID, title, due, tags)
				return nil, err
			}),
			slop.ActionOpts{
				Label:    "Edit task",
				Estimate: "instant",
				Params:   map[string]string{"title": "string", "due": "string", "tags": "string"},
			},
		)
	}

	actions["delete"] = slop.WithOpts(
		slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
			return nil, store.Delete(taskID)
		}),
		slop.ActionOpts{Label: "Delete task", Dangerous: true, Estimate: "instant"},
	)

	actions["read_notes"] = slop.WithOpts(
		slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
			found, err := store.Find(taskID)
			if err != nil {
				return nil, err
			}
			return map[string]any{"content": found.Notes}, nil
		}),
		slop.ActionOpts{
			Label:       "Read full notes",
			Description: "Fetch the complete notes for this task",
			Idempotent:  true,
			Estimate:    "instant",
		},
	)

	actions["write_notes"] = slop.WithOpts(
		slop.HandlerFunc(func(ctx context.Context, p slop.Params) (any, error) {
			content := p.String("content")
			_, err := store.SetNotes(taskID, content)
			return nil, err
		}),
		slop.ActionOpts{
			Label:    "Write notes",
			Estimate: "instant",
			Params:   map[string]string{"content": "string"},
		},
	)

	meta := &slop.Meta{
		Salience: floatPtr(sal),
	}
	if urg != "" {
		meta.Urgency = urg
	}
	if reason != "" {
		meta.Reason = reason
	}

	return slop.Item{
		ID:      t.ID,
		Props:   props,
		Actions: actions,
		Meta:    meta,
	}
}

func buildTagsNode(store *Store) slop.Node {
	tasks, err := store.All()
	if err != nil {
		return slop.Node{Type: "collection", Summary: "error loading tags"}
	}

	tagCounts := map[string]int{}
	for _, t := range tasks {
		for _, tag := range t.Tags {
			tagCounts[tag]++
		}
	}

	// Build summary
	type tagCount struct {
		name  string
		count int
	}
	sorted := make([]tagCount, 0, len(tagCounts))
	for name, count := range tagCounts {
		sorted = append(sorted, tagCount{name, count})
	}
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].count > sorted[j].count })

	parts := make([]string, len(sorted))
	for i, tc := range sorted {
		parts[i] = fmt.Sprintf("%s (%d)", tc.name, tc.count)
	}

	return slop.Node{
		Type:    "collection",
		Props:   slop.Props{"count": len(tagCounts)},
		Summary: fmt.Sprintf("%d tags: %s", len(tagCounts), strings.Join(parts, ", ")),
	}
}

File: examples/cli/rust/src/provider.rs
use std::collections::HashMap;
use std::fs;
use std::sync::{Arc, Mutex};

use chrono::Utc;
use serde_json::{json, Value};

use slop_ai::{ActionOptions, SlopServer};

use crate::store::{compute_salience, content_ref_for, Store, Task};

const WINDOW_SIZE: usize = 25;

pub fn setup_provider(store: Arc<Mutex<Store>>) -> SlopServer {
    let slop = SlopServer::new("tsk", "tsk");

    // --- Search action on tasks collection ---
    {
        let s = store.clone();
        slop.action_with(
            "tasks",
            "search",
            move |params: &Value| {
                let query = params["query"].as_str().unwrap_or("");
                let tasks = s.lock().unwrap().load();
                let q = query.to_lowercase();
                let matches: Vec<Value> = tasks
                    .iter()
                    .filter(|t| {
                        t.title.to_lowercase().contains(&q)
                            || t.tags.iter().any(|tag| tag.to_lowercase().contains(&q))
                    })
                    .map(|t| {
                        json!({
                            "id": t.id,
                            "title": &t.title,
                            "done": t.done,
                            "tags": &t.tags,
                        })
                    })
                    .collect();
                Ok(Some(json!({ "results": matches, "count": matches.len() })))
            },
            ActionOptions::new()
                .label("Search tasks")
                .description("Search tasks by title or tag")
                .idempotent(true)
                .estimate("instant")
                .params(json!({
                    "type": "object",
                    "properties": {
                        "query": { "type": "string", "description": "Search term (matches title and tags)" }
                    },
                    "required": ["query"]
                })),
        );
    }

    // --- Static: user context ---
    {
        let s = store.clone();
        slop.register_fn("user", move || {
            let st = s.lock().unwrap();
            let tasks = st.load();
            let total = tasks.len();
            let done = tasks.iter().filter(|t| t.done).count();
            json!({
                "type": "context",
                "props": {
                    "file": st.path.display().to_string(),
                    "total_tasks": total,
                    "total_done": done,
                },
            })
        });
    }

    // --- Dynamic: tasks collection ---
    {
        let s = store.clone();
        slop.register_fn("tasks", move || {
            let st = s.lock().unwrap();
            let tasks = st.load();
            build_tasks_descriptor(&tasks)
        });
    }

    // --- Dynamic: tags collection ---
    {
        let s = store.clone();
        slop.register_fn("tags", move || {
            let st = s.lock().unwrap();
            let tasks = st.load();
            build_tags_descriptor(&tasks)
        });
    }

    // --- Task actions on the collection ---
    register_collection_actions(&slop, store.clone());

    // --- Per-item actions (registered at "tasks" path, but handled via item path) ---
    register_item_actions(&slop, store.clone());

    // --- Tags actions ---
    {
        let s = store.clone();
        slop.action_with(
            "tags",
            "rename",
            move |params: &Value| {
                let old = params["old"].as_str().unwrap_or("");
                let new = params["new"].as_str().unwrap_or("");
                if old.is_empty() || new.is_empty() {
                    return Err(slop_ai::SlopError::ActionFailed {
                        code: "invalid_params".into(),
                        message: "old and new are required".into(),
                    });
                }
                let st = s.lock().unwrap();
                let mut tasks = st.load();
                let mut count = 0;
                for task in &mut tasks {
                    if let Some(pos) = task.tags.iter().position(|t| t == old) {
                        task.tags[pos] = new.to_string();
                        count += 1;
                    }
                }
                st.save(&tasks);
                Ok(Some(json!({ "renamed": count })))
            },
            ActionOptions::new()
                .label("Rename tag")
                .params(json!({
                    "type": "object",
                    "properties": {
                        "old": { "type": "string" },
                        "new": { "type": "string" }
                    },
                    "required": ["old", "new"]
                })),
        );
    }

    slop
}

fn register_collection_actions(slop: &SlopServer, store: Arc<Mutex<Store>>) {
    // add
    {
        let s = store.clone();
        slop.action_with(
            "tasks",
            "add",
            move |params: &Value| {
                let title = params["title"]
                    .as_str()
                    .ok_or_else(|| slop_ai::SlopError::ActionFailed {
                        code: "invalid_params".into(),
                        message: "title is required".into(),
                    })?;
                let st = s.lock().unwrap();
                let mut tasks = st.load();
                let id = st.next_id(&tasks);

                let due = params["due"].as_str().map(|d| resolve_date(d));
                let tags: Vec<String> = params["tags"]
                    .as_str()
                    .map(|t| t.split(',').map(|s| s.trim().to_string()).collect())
                    .unwrap_or_default();

                tasks.push(Task {
                    id: id.clone(),
                    title: title.to_string(),
                    done: false,
                    due,
                    tags,
                    notes: String::new(),
                    created: Utc::now().to_rfc3339(),
                    completed_at: None,
                });
                st.save(&tasks);
                Ok(Some(json!({ "id": id })))
            },
            ActionOptions::new()
                .label("Add task")
                .estimate("instant")
                .params(json!({
                    "type": "object",
                    "properties": {
                        "title": { "type": "string" },
                        "due": { "type": "string", "description": "ISO date or relative: 'today', 'tomorrow'" },
                        "tags": { "type": "string", "description": "Comma-separated tags" }
                    },
                    "required": ["title"]
                })),
        );
    }

    // clear_done
    {
        let s = store.clone();
        slop.action_with(
            "tasks",
            "clear_done",
            move |_params: &Value| {
                let st = s.lock().unwrap();
                let mut tasks = st.load();
                let before = tasks.len();
                tasks.retain(|t| !t.done);
                let removed = before - tasks.len();
                st.save(&tasks);
                Ok(Some(json!({ "removed": removed })))
            },
            ActionOptions::new()
                .label("Clear completed")
                .dangerous(true)
                .estimate("instant"),
        );
    }

    // export (async-style)
    {
        let s = store.clone();
        slop.action_with(
            "tasks",
            "export",
            move |params: &Value| {
                let format = params["format"].as_str().unwrap_or("json");
                let st = s.lock().unwrap();
                let tasks = st.load();

                let content = match format {
                    "csv" => {
                        let mut out = String::from("id,title,done,due,tags\n");
                        for t in &tasks {
                            let tags = t.tags.join(";");
                            let due = t.due.as_deref().unwrap_or("");
                            out.push_str(&format!("{},{},{},{},{}\n", t.id, t.title, t.done, due, tags));
                        }
                        out
                    }
                    "markdown" => {
                        let mut out = String::from("# Tasks\n\n");
                        for t in &tasks {
                            let check = if t.done { "x" } else { " " };
                            let due = t.due.as_deref().map(|d| format!(" (due: {d})")).unwrap_or_default();
                            out.push_str(&format!("- [{check}] {}{due}\n", t.title));
                        }
                        out
                    }
                    _ => serde_json::to_string_pretty(&tasks).unwrap_or_default(),
                };

                Ok(Some(json!({
                    "format": format,
                    "content": content,
                    "task_count": tasks.len(),
                })))
            },
            ActionOptions::new()
                .label("Export tasks")
                .description("Export tasks to a file")
                .estimate("slow")
                .params(json!({
                    "type": "object",
                    "properties": {
                        "format": { "type": "string", "enum": ["json", "csv", "markdown"] }
                    },
                    "required": ["format"]
                })),
        );
    }
}

fn register_item_actions(slop: &SlopServer, store: Arc<Mutex<Store>>) {
    // Register all action variants for every known task.
    // We register both done/undo for every task; the descriptor controls which
    // affordances are visible, and the handler works regardless of current state.
    // After an "add" action creates a new task, the auto-refresh rebuilds the tree
    // and the new task's affordances come from the descriptor. Its handlers are
    // registered here at setup time only for existing tasks, but new tasks get
    // handlers registered by the "add" action handler.
    let tasks = store.lock().unwrap().load();
    for task in &tasks {
        register_single_task_actions(slop, &store, &task.id);
    }

    // Also patch the "add" handler to register item actions for newly created tasks.
    // Since we can't easily do that, we instead pre-register a generous range of IDs.
    // A more robust approach: register handlers lazily. But for this example, we
    // register up to t-100 to cover new tasks.
    let max_id = tasks
        .iter()
        .filter_map(|t| t.id.strip_prefix("t-").and_then(|n| n.parse::<u32>().ok()))
        .max()
        .unwrap_or(0);
    for i in (max_id + 1)..=(max_id + 100) {
        let tid = format!("t-{i}");
        register_single_task_actions(slop, &store, &tid);
    }
}

fn register_single_task_actions(slop: &SlopServer, store: &Arc<Mutex<Store>>, task_id: &str) {
    let task_path = format!("tasks/{}", task_id);

    // done
    {
        let s = store.clone();
        let tid = task_id.to_string();
        slop.action(&task_path, "done", move |_params: &Value| {
            let st = s.lock().unwrap();
            let mut tasks = st.load();
            if let Some(t) = tasks.iter_mut().find(|t| t.id == tid) {
                t.done = true;
                t.completed_at = Some(Utc::now().to_rfc3339());
                st.save(&tasks);
                Ok(Some(json!({ "id": tid })))
            } else {
                Err(slop_ai::SlopError::ActionFailed {
                    code: "not_found".into(),
                    message: format!("Task {} not found", tid),
                })
            }
        });
    }

    // undo
    {
        let s = store.clone();
        let tid = task_id.to_string();
        slop.action(&task_path, "undo", move |_params: &Value| {
            let st = s.lock().unwrap();
            let mut tasks = st.load();
            if let Some(t) = tasks.iter_mut().find(|t| t.id == tid) {
                t.done = false;
                t.completed_at = None;
                st.save(&tasks);
                Ok(Some(json!({ "id": tid })))
            } else {
                Err(slop_ai::SlopError::ActionFailed {
                    code: "not_found".into(),
                    message: format!("Task {} not found", tid),
                })
            }
        });
    }

    // edit
    {
        let s = store.clone();
        let tid = task_id.to_string();
        slop.action(&task_path, "edit", move |params: &Value| {
            let st = s.lock().unwrap();
            let mut tasks = st.load();
            if let Some(t) = tasks.iter_mut().find(|t| t.id == tid) {
                if let Some(title) = params["title"].as_str() {
                    t.title = title.to_string();
                }
                if let Some(due) = params["due"].as_str() {
                    t.due = Some(resolve_date(due));
                }
                if let Some(tags) = params["tags"].as_str() {
                    t.tags = tags.split(',').map(|s| s.trim().to_string()).collect();
                }
                st.save(&tasks);
                Ok(Some(json!({ "id": tid })))
            } else {
                Err(slop_ai::SlopError::ActionFailed {
                    code: "not_found".into(),
                    message: format!("Task {} not found", tid),
                })
            }
        });
    }

    // delete
    {
        let s = store.clone();
        let tid = task_id.to_string();
        slop.action(&task_path, "delete", move |_params: &Value| {
            let st = s.lock().unwrap();
            let mut tasks = st.load();
            let len_before = tasks.len();
            tasks.retain(|t| t.id != tid);
            if tasks.len() < len_before {
                st.save(&tasks);
                Ok(Some(json!({ "id": tid })))
            } else {
                Err(slop_ai::SlopError::ActionFailed {
                    code: "not_found".into(),
                    message: format!("Task {} not found", tid),
                })
            }
        });
    }

    // read_notes
    {
        let s = store.clone();
        let tid = task_id.to_string();
        slop.action(&task_path, "read_notes", move |_params: &Value| {
            let st = s.lock().unwrap();
            let tasks = st.load();
            if let Some(t) = tasks.iter().find(|t| t.id == tid) {
                Ok(Some(json!({ "content": &t.notes })))
            } else {
                Ok(Some(json!({ "content": "" })))
            }
        });
    }

    // write_notes
    {
        let s = store.clone();
        let tid = task_id.to_string();
        slop.action(&task_path, "write_notes", move |params: &Value| {
            let content = params["content"].as_str().unwrap_or("");
            let st = s.lock().unwrap();
            let mut tasks = st.load();
            if let Some(t) = tasks.iter_mut().find(|t| t.id == tid) {
                t.notes = content.to_string();
                st.save(&tasks);
                Ok(Some(json!({ "id": tid })))
            } else {
                Err(slop_ai::SlopError::ActionFailed {
                    code: "not_found".into(),
                    message: format!("Task {} not found", tid),
                })
            }
        });
    }
}

fn build_tasks_descriptor(tasks: &[Task]) -> Value {
    let total = tasks.len();
    let pending = tasks.iter().filter(|t| !t.done).count();
    let done_count = total - pending;

    // Sort by salience
    let mut sorted: Vec<&Task> = tasks.iter().collect();
    sorted.sort_by(|a, b| {
        let (sa, _, _) = compute_salience(a);
        let (sb, _, _) = compute_salience(b);
        sb.partial_cmp(&sa).unwrap_or(std::cmp::Ordering::Equal)
    });

    let overdue = sorted
        .iter()
        .filter(|t| {
            let (s, _, _) = compute_salience(t);
            s >= 1.0 && !t.done
        })
        .count();

    let window_count = total.min(WINDOW_SIZE);
    let windowed: Vec<&Task> = sorted.iter().take(window_count).copied().collect();

    let items: Vec<Value> = windowed
        .iter()
        .map(|t| {
            let (salience, urgency, reason) = compute_salience(t);

            let mut meta = json!({ "salience": salience });
            if let Some(u) = &urgency {
                meta["urgency"] = json!(u);
            }
            if let Some(r) = &reason {
                meta["reason"] = json!(r);
            }

            let mut props = json!({
                "title": &t.title,
                "done": t.done,
                "tags": &t.tags,
            });
            if let Some(due) = &t.due {
                props["due"] = json!(due);
            }
            if let Some(cat) = &t.completed_at {
                props["completed_at"] = json!(cat);
            }

            // Build actions for this item
            let actions = if t.done {
                json!({
                    "undo": { "label": "Mark incomplete", "estimate": "instant" },
                    "delete": { "label": "Delete task", "dangerous": true, "estimate": "instant" },
                })
            } else {
                json!({
                    "done": { "label": "Complete task", "estimate": "instant" },
                    "edit": {
                        "label": "Edit task",
                        "estimate": "instant",
                        "params": {
                            "type": "object",
                            "properties": {
                                "title": { "type": "string" },
                                "due": { "type": "string" },
                                "tags": { "type": "string" }
                            }
                        }
                    },
                    "delete": { "label": "Delete task", "dangerous": true, "estimate": "instant" },
                    "read_notes": {
                        "label": "Read full notes",
                        "description": "Fetch the complete notes for this task",
                        "idempotent": true,
                        "estimate": "instant"
                    },
                    "write_notes": {
                        "label": "Write notes",
                        "estimate": "instant",
                        "params": {
                            "type": "object",
                            "properties": {
                                "content": { "type": "string" }
                            },
                            "required": ["content"]
                        }
                    },
                })
            };

            // Embed content_ref inside props since normalize_item reads props directly
            if !t.done {
                props["content_ref"] = content_ref_for(t);
            }

            json!({
                "id": &t.id,
                "props": props,
                "meta": meta,
                "actions": actions,
            })
        })
        .collect();

    let summary = format!(
        "{total} tasks: {pending} pending, {done_count} done, {overdue} overdue"
    );

    json!({
        "type": "collection",
        "props": {
            "count": total,
            "pending": pending,
            "overdue": overdue,
        },
        "summary": summary,
        "meta": {
            "total_children": total,
        },
        "window": {
            "items": items,
            "total": total,
            "offset": 0,
        },
        "actions": {
            "add": {
                "label": "Add task",
                "estimate": "instant",
                "params": {
                    "type": "object",
                    "properties": {
                        "title": { "type": "string" },
                        "due": { "type": "string", "description": "ISO date or relative" },
                        "tags": { "type": "string", "description": "Comma-separated tags" }
                    },
                    "required": ["title"]
                }
            },
            "clear_done": {
                "label": "Clear completed",
                "description": "Remove all completed tasks",
                "dangerous": true,
                "estimate": "instant"
            },
            "export": {
                "label": "Export tasks",
                "description": "Export tasks to a file",
                "estimate": "slow",
                "params": {
                    "type": "object",
                    "properties": {
                        "format": { "type": "string", "enum": ["json", "csv", "markdown"] }
                    },
                    "required": ["format"]
                }
            },
            "search": {
                "label": "Search tasks",
                "description": "Search tasks by title or tag",
                "idempotent": true,
                "estimate": "instant",
                "params": {
                    "type": "object",
                    "properties": {
                        "query": { "type": "string", "description": "Search term" }
                    },
                    "required": ["query"]
                }
            }
        },
    })
}

fn build_tags_descriptor(tasks: &[Task]) -> Value {
    let mut tag_counts: HashMap<String, usize> = HashMap::new();
    for task in tasks {
        for tag in &task.tags {
            *tag_counts.entry(tag.clone()).or_default() += 1;
        }
    }

    let mut sorted_tags: Vec<(&String, &usize)> = tag_counts.iter().collect();
    sorted_tags.sort_by(|a, b| b.1.cmp(a.1));

    let tag_summary = sorted_tags
        .iter()
        .map(|(name, count)| format!("{name} ({count})"))
        .collect::<Vec<_>>()
        .join(", ");

    let count = tag_counts.len();

    json!({
        "type": "collection",
        "props": { "count": count },
        "summary": format!("{count} tags: {tag_summary}"),
    })
}

fn resolve_date(input: &str) -> String {
    let today = chrono::Utc::now().date_naive();
    match input.to_lowercase().as_str() {
        "today" => today.format("%Y-%m-%d").to_string(),
        "tomorrow" => (today + chrono::Duration::days(1))
            .format("%Y-%m-%d")
            .to_string(),
        _ => input.to_string(),
    }
}

/// Write the provider discovery file to ~/.slop/providers/tsk.json.
pub fn write_discovery(store: &Arc<Mutex<Store>>, socket_path: &str) {
    let home = match dirs::home_dir() {
        Some(h) => h,
        None => return,
    };
    let dir = home.join(".slop").join("providers");
    let _ = fs::create_dir_all(&dir);
    let path = dir.join("tsk.json");

    let st = store.lock().unwrap();
    let tasks = st.load();
    let total = tasks.len();
    let pending = tasks.iter().filter(|t| !t.done).count();
    let overdue = tasks
        .iter()
        .filter(|t| {
            let (s, _, _) = compute_salience(t);
            s >= 1.0 && !t.done
        })
        .count();

    let desc = json!({
        "id": "tsk",
        "name": "tsk",
        "version": "0.1.0",
        "slop_version": "0.1",
        "transport": { "type": "unix", "path": socket_path },
        "pid": std::process::id(),
        "capabilities": ["state", "patches", "affordances", "attention"],
        "description": format!("Task manager with {total} tasks ({pending} pending, {overdue} overdue)")
    });

    let _ = fs::write(&path, serde_json::to_string_pretty(&desc).unwrap());
}

/// Remove the discovery file.
pub fn remove_discovery() {
    let home = match dirs::home_dir() {
        Some(h) => h,
        None => return,
    };
    let path = home.join(".slop").join("providers").join("tsk.json");
    let _ = fs::remove_file(path);
}
